waku/incentivization/txid_proof.nim

```diff
@@ -0,0 +1,76 @@
+import std/options, chronos, web3, stew/byteutils, stint, results, chronicles
+
+import waku/incentivization/rpc
+
+const SimpleTransferGasUsed = Quantity(21000)
+
+proc eth_getTransactionByHash_async(
+    txHash: TxHash, web3: Web3
+): Future[TransactionObject] {.async.} =
+  await web3.provider.eth_getTransactionByHash(txHash)
+
+proc getMinedTransactionReceipt_async(
+    txHash: TxHash, web3: Web3
+): Future[ReceiptObject] {.async.} =
+  await web3.getMinedTransactionReceipt(txHash)
+
+proc getTxAndTxReceipt(
+    txHash: TxHash, web3: Web3
+): Future[(TransactionObject, ReceiptObject)] {.async.} =
+  let txFuture = eth_getTransactionByHash_async(txHash, web3)
+  let receiptFuture = getMinedTransactionReceipt_async(txHash, web3)
+  let tx = await txFuture
+  let txReceipt = await receiptFuture
+  return (tx, txReceipt)
```

Comment:
I think this way is a bit more clear that it's being done in parallel

```suggestion
proc getTxAndTxReceipt(
    txHash: TxHash, web3: Web3
): Future[(TransactionObject, ReceiptObject)] {.async.} =
  let txFuture = eth_getTransactionByHash_async(txHash, web3)
  let receiptFuture = getMinedTransactionReceipt_async(txHash, web3)
  await allFutures(txFuture, receiptFuture)
  return (txFuture.read(), receiptFuture.read())

```

Reply:
Hm, when I applied this suggestion, the [test](https://github.com/waku-org/nwaku/blob/d6ac598eb58b9358769f9c438e1003184c455617/tests/incentivization/test_poc.nim#L37) that checks that a non-existent transaction is not eligible fell into an infinite loop. It seems that the updated code tries to query the zero hash again and again, but the transaction with that hash is never there (as expected). What should be the proper handling of this scenario in terms of async operations?

Reply:
Ooh I just reproduced and saw the issue, apologies!

So my understanding is that `eth_getTransactionByHash_async` throws an exception and if we don't await for it we won't catch it immediately.

On the other hand, `getMinedTransactionReceipt_async` keeps retrying if it fails to get the receipt, so in this case we are awaiting for both futures to complete but `receiptFuture` never completes because it keeps retrying indefinitely.

I think we need to add a timeout in `getMinedTransactionReceipt_async` and don't allow it to keep retrying without ever stopping.`getMinedTransactionReceipt_async` can leave you hanging indefinitely because of the future not timing out, this is a issue you will potentially find in other usages of the proc - in this particular test we were lucky that the exception in the other procedure stopped it.

Anyway, in summary - current solution that passes the test also works, but be aware that `getMinedTransactionReceipt_async` not timing out is an issue that can leave you in an infinite loop in cases beyond this test. If by design you do want that behavior of infinite retries, then current solution works perfectly and the change I proposed is not relevant :) if you do want to add a timeout to `getMinedTransactionReceipt_async`, then the proposed solution should work once the timeout is in place

Reply:
Looks amazing!Thanks so so much! 😍